src/profiler.rs

```diff
@@ -465,6 +509,74 @@ impl Profiler {
     }
 }
 
+// The kernel thread ID ( the same returned by a call to gettid(2))
+// is not the same thing as the thread ID returned by pthread_self().
+//
+// Here we fetch a list of kernel thread IDs from /proc/{pid}/task
+#[cfg(target_os = "linux")]
+fn get_tids(pid_t: u32) -> Vec<u32> {
+    std::fs::read_dir(format!("/proc/{}/task", pid_t))
+        .unwrap()
+        .into_iter()
+        .filter_map(|entry| entry.map_or(None, |entry| entry.file_name().into_string().ok()))
+        .filter_map(|tid| tid.parse::<u32>().ok())
+        .collect()
+}
+
+#[cfg(target_os = "linux")]
+fn fire_rt_signals(running: Arc<AtomicBool>, frequency: u64) {
+    use std::{thread, time::Duration};
+
+    let pid = nix::unistd::Pid::this().as_raw() as u32;
+    let prof_tid = unsafe { libc::syscall(libc::SYS_gettid) as u32 };
+
+    let interval = 1e6 as u64 / frequency;
+    let tv_usec = interval % 1e6 as u64;
+
+    while running.load(Ordering::SeqCst) {
+        let tids = get_tids(pid);
+        for tid in tids {
```

Comment:
Maybe you would want to check for the `running` flag on each iteration. As it is right now, the signalling thread will always finish its queue of threads when you want to stop the profiling. Whereas I would rather assume the intention is to *stop* right away, without collecting a few more runaway samples after the call to `stop`.

Reply:
I think here there is a performance trade-off on both sides to be made.

I'm not sure if we should check the condition for each `tid` at each interval considering we're using an `Arc<AtomicBool>`.

On the other side it's true that we're finishing the remaining of the queue even if it's stopped, but that only happens once (when stopping and not at every interval) and considering that when we stop the profiler the `signal_handler` gets "unregistered", those signals won't be handled and hence we won't collect "extra" samples.